Class/Warrior/Spells/SunderArmor.cpp

```diff
@@ -0,0 +1,81 @@
+#include "SunderArmor.h"
+
+#include "Buff.h"
+#include "CastComplete.h"
+#include "CharacterStats.h"
+#include "CombatRoll.h"
+#include "CooldownControl.h"
+#include "HeroicStrike.h"
+#include "MainhandAttack.h"
+#include "OffhandAttack.h"
+#include "Warrior.h"
+#include "WarriorSpells.h"
+
+SunderArmor::SunderArmor(Warrior* pchar, WarriorSpells* spells) :
+    // TODO Find the right spell icon
+    Spell("SunderArmor", "Assets/ability/Ability_warrior_decisivestrike.png",
+          pchar, new CooldownControl(pchar, 0.0), RestrictedByGcd::Yes,
+          ResourceType::Rage,
+          15),
+    TalentRequirer(QVector<TalentRequirerInfo*>{new TalentRequirerInfo("Improved Sunder Armor", 3, DisabledAtZero::No)}),
+    warr(pchar),
+    spells(spells),
+    // TODO find these lower level values
+    innate_threat_by_spell_rank({0, 0, 0, 0, 260 }),
+    innate_threat(260),
+    talent_ranks({15, 14, 13, 12})
+{
+}
+
+SunderArmor::~SunderArmor() {
+    delete cooldown;
+}
+
+void SunderArmor::increase_talent_rank_effect(const QString&, const int curr) {
+    resource_cost = talent_ranks[curr];
+}
+
+void SunderArmor::decrease_talent_rank_effect(const QString&, const int curr) {
+    resource_cost = talent_ranks[curr];
+}
+
+SpellStatus SunderArmor::is_ready_spell_specific() const {
+    if (warr->in_battle_stance())
+      return SpellStatus::InBattleStance;
+
+    if (warr->in_berserker_stance())
+      return SpellStatus::InBerserkerStance;
+
+    return warr->on_stance_cooldown() ? SpellStatus::OnStanceCooldown : SpellStatus::Available;
+}
+
+void SunderArmor::spell_effect() {
+    // Sunder armor can't crit
+    const int result = roll->get_melee_ability_result(warr->get_mh_wpn_skill(), 0);
+
+    cooldown->add_gcd_event();
+
+    if (result == PhysicalAttackResult::MISS) {
+        increment_miss();
+        warr->lose_rage(static_cast<unsigned>(round(warr->get_resource_level(resource_type) * 0.16)));
```

Comment:
For almost (or is it all?) other warrior abilities MISS means all the rage is lost, whereas *some* rage is returned for DODGE and PARRY. Are you certain MISS should return rage for SA? Remove the multiplication if it should not return any rage.

Reply:
Nope, I just copied from execute. Guess it was a bad choice. I'll take the values from MS.

Reply:
Execute is (or was at least in the beta, see [the commit](https://github.com/timhul/ClassicSim/commit/49fed2efb3f7dc88b7a8280640b20571e7e6df30#diff-eba1142a21a88ded1df7803247b8d7af)) a special case for resource cost, although recently this has been debated in https://github.com/timhul/ClassicSim/issues/123.